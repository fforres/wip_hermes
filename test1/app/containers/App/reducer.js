/*
 * AppReducer
 *
 * The reducer takes care of our data. Using actions, we can change our
 * application state.
 * To add a new action, add it to the switch statement in the reducer function
 *
 * Example:
 * case YOUR_ACTION_CONSTANT:
 *   return state.set('yourStateVariable', true);
 */

import { combineReducers } from 'redux-immutable';
import { List } from 'immutable';

import {
  LOAD_REPOS_SUCCESS,
  LOAD_REPOS,
  LOAD_REPOS_ERROR,
} from './constants';

function loading(state = false, action) {
  switch (action.type) {
    case LOAD_REPOS:
      return true;
    case LOAD_REPOS_SUCCESS:
      return false;
    case LOAD_REPOS_ERROR:
      return false;
    default:
      return state;
  }
}

function error(state = false, action) {
  switch (action.type) {
    case LOAD_REPOS:
      return false;
    case LOAD_REPOS_SUCCESS:
      return false;
    case LOAD_REPOS_ERROR:
      return action.error;
    default:
      return state;
  }
}

function currentUser(state = false, action) {
  switch (action.type) {
    case LOAD_REPOS_SUCCESS:
      return action.username;
    default:
      return state;
  }
}

function repositories(state = false, action) {
  if (List.isList(state)) {
    // normalize the state from the server
    state = state.toJS(); // eslint-disable-line no-param-reassign
  }

  switch (action.type) {
    case LOAD_REPOS:
      return false;
    case LOAD_REPOS_SUCCESS:
      return action.repos;
    default:
      return state;
  }
}

const userData = combineReducers({
  repositories,
});

export default combineReducers({
  loading,
  error,
  currentUser,
  userData,
});
