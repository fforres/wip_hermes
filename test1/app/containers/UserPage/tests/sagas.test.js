/**
 * Tests for UserPage sagas
 */

import { take, put, cancel, takeLatest } from 'redux-saga/effects';
import { createMockTask } from 'redux-saga/utils';
import { LOCATION_CHANGE } from 'react-router-redux';

import { LOAD_REPOS } from 'containers/App/constants';
import { reposLoaded, repoLoadingError } from 'containers/App/actions';

import { getRepos, githubData } from '../sagas';

const username = 'mxstbr';

/* eslint-disable redux-saga/yield-effects */
describe('getRepos Saga', () => {
  let getReposGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getReposGenerator = getRepos({ username });

    const callDescriptor = getReposGenerator.next(username).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the reposLoaded action if it requests the data successfully', () => {
    const response = [{
      name: 'First repo',
    }, {
      name: 'Second repo',
    }];
    const putDescriptor = getReposGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(reposLoaded(response, username)));
  });

  it('should call the repoLoadingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getReposGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(repoLoadingError(response)));
  });
});

describe('githubDataSaga Saga', () => {
  let generator;
  let taskMock;

  beforeEach(() => {
    generator = githubData();
    taskMock = createMockTask();
  });

  it('should start task to watch for LOAD_REPOS action', () => {
    const takeLatestDescriptor = generator.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_REPOS, getRepos));
  });

  it('should yield until LOCATION_CHANGE action', () => {
    generator.next();
    const takeDescriptor = generator.next();
    expect(takeDescriptor.value).toEqual(take(LOCATION_CHANGE));
  });

  it('should cancel after LOCATION_CHANGE action', () => {
    generator.next();
    generator.next(taskMock);
    const cancelDescriptor = generator.next();
    expect(cancelDescriptor.value).toEqual(cancel(taskMock));
  });
});
