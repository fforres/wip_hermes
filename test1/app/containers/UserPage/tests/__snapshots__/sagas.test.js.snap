exports[`getRepos Saga should call the repoLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://api.github.com/users/mxstbr/repos?type=all&sort=updated",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getRepos Saga should dispatch the reposLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://api.github.com/users/mxstbr/repos?type=all&sort=updated",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
